EIPS/eip-5843.md

```diff
@@ -0,0 +1,437 @@
+* It takes significant development/testing time and effort to ensure performance and correctness.  And even this effort is often not sufficient as precompiles have been the source of previous consensus bugs (e.g.: bn128 ecmul not doing subgroup checks, ecrecover bug from 2016, Nimbus modexp input padding bug, others?).
+* The skillset required to implement performant crypto software is very specialized.  This makes developing and maintaining multiple production-grade implementations of the same specification difficult.
+    * This can have the effect that multiple clients will default to use the same implementation.  It can introduce a potential single point of failure for the network by effectively enshrining specific implementations into the spec if the pricing model is aggressive.
+* New research developments can make existing precompiles outdated and necessitate new ones.  Removing old precompiles completely will likely be impossible (the current situation regarding `SELFDESTRUCT` removal illustrates the difficulties of removing old functionality).
+
+To reduce pressure for new crypto precompiles, we can identify and optimize operations which form a common bottleneck for many desired use-cases in order to make them efficiently-implementable as EVM contracts.  Aformentioned elliptic curve precompiles share a common performance bottleneck: modular arithmetic (and specifically: modular multiplication) over a fixed, odd, often-prime modulus.
+
+Currently in the EVM, modular addition and multiplication are performed with the `ADDMOD` and `MULMOD` opcodes.  Inputs and a modulus are passed as stack items and a result is returned via the stack.  The opcodes are permissive: the modulus can be any number and inputs are not required to be less than the modulus.
+
+`ADDMOD` and `MULMOD` have a lot of utility.  However, they are restricted to values that fit within 256 bits.  They also cannot take advantage of optimizations that can greatly improve performance in many cases.
+
+If we add the requirement that inputs must be less than the modulus, modular addition and subtraction can be implemented efficiently with each requiring one extended-precision addition and one extended-precision subtraction. Normal modular multiplication is slower and requires an extended-precision multiplication to compute a double-width product, and a modular reduction.
+
+To improve the performance of modular multiplication, values can be expressed in Montgomery form. This allows for the use of Montgomery modular multiplication which replaces the division by the modulus with a bit-shift and single subtraction in the reduction step.
+
+Operating in Montgomery form requires precomputation of a modulus-specific constant as well as the cost of converting values in canonical form to/from Montgomery form. However, these costs are greatly offset by the savings gained when performing many modular multiplications.
```

Comment:
Reading to this, I feel the current title might be too general, consider update title to be more specific: `EVM Montgomery Modulus Instructions`?

Reply:
Thanks for the suggestion but I'd prefer to keep as-is for now.  The name EVMMAX is used a lot in the related implementation/toolset code that comes along with this EIP.